File: clients/python/baml_lib/_impl/deserializer/raw_wrapper/list_wrapper.py
import typing
from .raw_wrapper import RawWrapper
import json

T = typing.TypeVar("T", bound=typing.Any)


def filter_empty(
    x: typing.Iterable[typing.Optional[T]],
) -> typing.Iterable[T]:
    return typing.cast(typing.Iterable[T], filter(lambda v: v is not None, x))


@typing.final
class ListRawWrapper(RawWrapper):
    def __init__(self, val: typing.List[RawWrapper]) -> None:
        self.__val = val

    def as_self(self) -> typing.Optional[typing.Any]:
        return [item.as_self() for item in self.__val]

    def as_str(self, inner: bool) -> typing.Optional[str]:
        return json.dumps(self.as_self())

    def as_smart_str(self, inner: bool) -> typing.Optional[str]:
        if len(self.__val) == 1:
            return self.__val[0].as_smart_str(inner)
        return self.as_str(True)

    def as_int(self) -> typing.Optional[int]:
        if len(self.__val) == 0:
            return None
        for item in filter_empty(map(lambda v: v.as_int(), self.__val)):
            return item
        return None

    def as_float(self) -> typing.Optional[float]:
        if len(self.__val) == 0:
            return None
        for item in filter_empty(map(lambda v: v.as_float(), self.__val)):
            return item
        return None

    def as_bool(self) -> typing.Optional[bool]:
        if len(self.__val) == 0:
            return None
        for item in filter_empty(map(lambda v: v.as_bool(), self.__val)):
            return item
        return None

    def as_list(self) -> typing.Iterable[RawWrapper]:
        return self.__val

    def as_dict(
        self,
    ) -> typing.ItemsView[typing.Optional[RawWrapper], RawWrapper]:
        return {None: self}.items()

File: clients/ts/src/baml_lib/deserializer/raw_wrapper/list_wrapper.ts
// Import the RawWrapper class
import { RawWrapper } from './raw_wrapper';

// TypeScript doesn't have a direct equivalent of Python's filter_empty function,
// so we'll create a utility function to filter out null or undefined values from an array
function filterEmpty<T>(array: Array<T | undefined>): Array<T> {
    return array.filter((item): item is T => item !== undefined);
}

// TypeScript equivalent of the ListRawWrapper class
class ListRawWrapper<T extends RawWrapper> extends RawWrapper {
    private __val: Array<T>;

    constructor(val: Array<T>) {
        super();
        this.__val = val;
    }

    as_self(): any {
        return this.__val.map(item => item.as_self());
    }

    as_str(inner: boolean): string | undefined {
        return JSON.stringify(this.as_self(), null, inner ? 0 : 2);
    }

    as_smart_str(inner: boolean): string | undefined {
        if (this.__val.length === 1) {
            return this.__val[0].as_smart_str(inner);
        }
        return this.as_str(true);
    }

    as_int(): number | undefined {
        if (this.__val.length === 0) {
            return undefined;
        }
        for (const item of filterEmpty(this.__val.map(v => v.as_int()))) {
            return item;
        }
        return undefined;
    }

    as_float(): number | undefined {
        if (this.__val.length === 0) {
            return undefined;
        }
        for (const item of filterEmpty(this.__val.map(v => v.as_float()))) {
            return item;
        }
        return undefined;
    }

    as_bool(): boolean | undefined {
        if (this.__val.length === 0) {
            return undefined;
        }
        for (const item of filterEmpty(this.__val.map(v => v.as_bool()))) {
            return item;
        }
        return undefined;
    }

    as_list(): Array<T> {
        return this.__val;
    }

    as_dict(): Map<RawWrapper | undefined, RawWrapper> {
        return new Map([[undefined, this]]);
    }
}

export { ListRawWrapper };